gh-132983: Add documentation for compression.zstd

emmatyping opened the pull request:
This PR adds docs for the `compression` and `compression.zstd` modules.

<!-- gh-issue-number: gh-132983 -->
* Issue: gh-132983
<!-- /gh-issue-number -->


<!-- readthedocs-preview cpython-previews start -->
----
📚 Documentation preview 📚: https://cpython-previews--133911.org.readthedocs.build/en/133911/library/compression.zstd.html

<!-- readthedocs-preview cpython-previews end -->

emmatyping committed:
Add documentation for compression.zstd

emmatyping committed:
Add examples

emmatyping committed:
Fix camelcase name references

emmatyping committed:
CParameter->CompressionParameter

AA-Turner commented on Doc/library/archiving.rst:
```diff
@@ -5,13 +5,15 @@ Data Compression and Archiving
 
 The modules described in this chapter support data compression with the zlib,
-gzip, bzip2 and lzma algorithms, and the creation of ZIP- and tar-format
+gzip, bzip2, zstd, and lzma algorithms, and the creation of ZIP- and tar-format
```
New algorithm at the end?
```suggestion
gzip, bzip2, lzma, and zstd algorithms, and the creation of ZIP- and tar-format
```

AA-Turner commented on Doc/library/compression.rst:
```diff
@@ -0,0 +1,25 @@
+The :mod:`!compression` package
+=================================
```
```suggestion
The :mod:`!compression` package
===============================
```

AA-Turner commented on Doc/library/compression.rst:
```diff
@@ -0,0 +1,25 @@
+    canonical import name for the respective top-level module. The existing
+    modules are not currently deprecated and will not be removed prior to Python
+    3.19, but users are encouraged to migrate to the new import names when
+    feasible.
```
```suggestion
.. attention::

   The :mod:`!compression` package is the new location for the data compression
   modules in the standard library, listed below. The existing modules are not
   deprecated and will not be removed before Python 3.19. The new ``compression.*``
   import names are encouraged for use where practicable.
```

AA-Turner commented on Doc/library/compression.rst:
```diff
@@ -0,0 +1,25 @@
+
+* :mod:`!compression.zlib` -- Re-exports :mod:`zlib`
+
+* :mod:`compression.zstd` -- Wrapper for the Zstandard compression library
```
```suggestion
* :mod:`!compression.bz2` -- Re-exports :mod:`bz2`
* :mod:`!compression.gzip` -- Re-exports :mod:`gzip`
* :mod:`!compression.lzma` -- Re-exports :mod:`lzma`
* :mod:`!compression.zlib` -- Re-exports :mod:`zlib`
* :mod:`compression.zstd` -- Wrapper for the Zstandard compression library
```

AA-Turner commented on Doc/library/compression.rst:
```diff
@@ -0,0 +1,25 @@
+
+* :mod:`compression.zstd` -- Wrapper for the Zstandard compression library
+
+
```
```suggestion
```

AA-Turner commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,768 @@
+Reading and writing compressed files
+------------------------------------
+
+.. function:: open(file, /, mode="rb", *, level=None, options=None, zstd_dict=None, encoding=None, errors=None, newline=None)
```
wrap lines
```suggestion
.. function:: open(file, /, mode='rb', *, level=None, options=None, \
                   zstd_dict=None, encoding=None, errors=None, newline=None)
```

AA-Turner commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,768 @@
+   behavior, and line ending(s).
+
+
+.. class:: ZstdFile(file, /, mode="r", *, level=None, options=None, zstd_dict=None)
```
```suggestion
.. class:: ZstdFile(file, /, mode='r', *, level=None, options=None, zstd_dict=None)
```

AA-Turner commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,768 @@
+
+      ``'rb'`` for reading and ``'wb'`` for writing.
+
+
```
```suggestion
```

AA-Turner commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,768 @@
+
+   .. attribute:: name
+
+      The zstd file name. Equivalent to the :attr:`~io.FileIO.name`
```
I think 'zstd' here is incorrect / should be avoided
```suggestion
      The name of the Zstandard file. Equivalent to the :attr:`~io.FileIO.name`
```

AA-Turner commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,768 @@
+
+            * -
+              - | Digested
+                | dictionary
```
Is the use of line-blocks (`|`) intentional?

emmatyping replied:
Not particularly no. This is based on the table in the pyzstd docs. I think these should probably be removed (I'll push a commit later to do so).

AA-Turner commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,768 @@
+      ratios but slower compression.
+
+      .. seealso::
+         :class:`Strategy`
```
```suggestion
      .. seealso:: :class:`Strategy`
```

AA-Turner commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,768 @@
+   .. note::
+
+      The values of attributes of :class:`Strategy` are not necessarily stable
+      between zstd versions. Only the ordering may be relied upon.
```
I think we should rephrase, there shouldn't be an expectation that the Python user is able to control their zstd version (e.g. Windows, or a linux redistribuition)

emmatyping replied:
Yeah this should be rephrased. I don't mean to imply that the user should change their zstd version, merely that across different versions (different Python versions/OS/distro/etc.) the values could theoretically not be stable.

emmatyping replied:
I tried modifying this a little. If it's still unclear I can try again.

AA-Turner commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,768 @@
+.. function:: get_frame_info(frame_buffer)
+
+   Retrieve a :class:`FrameInfo`, containing metadata about a Zstandard frame.
+   Frames contain metadata related to the compressed data they hold.
```
```suggestion
   Retrieve a :class:`FrameInfo` object containing metadata about a Zstandard
   frame. Frames contain metadata related to the compressed data they hold.
```

AA-Turner commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,768 @@
+   Frames contain metadata related to the compressed data they hold.
+
+
+.. class:: FrameInfo()
```
```suggestion
.. class:: FrameInfo
```

AA-Turner commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,768 @@
+
+.. attribute:: zstd_version_info
+
+   Version number of the runtime zstd library as a tuple of int
```
```suggestion
   Version number of the runtime zstd library as a tuple of integers
```

AA-Turner commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,768 @@
+--------
+
+Reading in a compressed file::
+
```
we prefer `code-block` for new docs
```suggestion
Reading in a compressed file:

.. code-block:: python
```

AA-Turner commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,768 @@
+       file_content = f.read()
+
+Creating a compressed file::
+
```
```suggestion
Creating a compressed file:

.. code-block:: python
```

AA-Turner commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,768 @@
+       f.write(data)
+
+Compressing data in memory::
+
```
```suggestion
Compressing data in memory:

.. code-block:: python
```

AA-Turner commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,768 @@
+   data_out = zstd.compress(data_in)
+
+Incremental compression::
+
```
```suggestion
Incremental compression:

.. code-block:: python
```

AA-Turner commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,768 @@
+   result = b"".join([out1, out2, out3, out4])
+
+Writing compressed data to an already-open file::
+
```
```suggestion
Writing compressed data to an already-open file:

.. code-block:: python
```

AA-Turner commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,768 @@
+       f.write(b"Not compressed\n")
+
+Creating a compressed file using compression parameters::
+
```
```suggestion
Creating a compressed file using compression parameters:

.. code-block:: python
```

AA-Turner reviewed (commented):
syntactic review

emmatyping committed:
Apply suggestions from AA-Turner

Co-authored-by: Adam Turner <9087854+AA-Turner@users.noreply.github.com>

StanFromIreland commented on Doc/library/compression.rst:
```diff
@@ -0,0 +1,18 @@
+The :mod:`!compression` package
+===============================
+
```
Can we have a source code link like all other library pages.

AA-Turner replied:
Not needed, the corollary is `concurrent`.

StanFromIreland commented on Doc/library/compression.rst:
```diff
@@ -0,0 +1,18 @@
+
+.. versionadded:: 3.14
+
+.. attention::
```
This does not need the directive IMO

AA-Turner replied:
I saw it as a transitional thing, but happy to remove or keep

emmatyping replied:
I think it is useful and important to draw attention to this notice which, as Adam said, is transitional. We want to make sure users are aware that these names are the new canonical names.

StanFromIreland replied:
![image](https://github.com/user-attachments/assets/2ba23866-ba8f-43cd-819b-bec4833d7844)

I don't have a strong preference against, but IMO I still see little value.

StanFromIreland commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,772 @@
+
+.. function:: open(file, /, mode="rb", *, level=None, options=None, zstd_dict=None, encoding=None, errors=None, newline=None)
+
+   Open an Zstandard-compressed file in binary or text mode, returning a
```
```suggestion
   Open a Zstandard-compressed file in binary or text mode, returning a
```
Grammar

AA-Turner replied:
we don't need the dash here either

StanFromIreland commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,772 @@
+Reading in a compressed file:
+
+.. code-block:: python
+   from compression import zstd
```
```suggestion

.. code-block:: python

   from compression import zstd
```
You need a gap between the code and directives.

StanFromIreland commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,772 @@
+Creating a compressed file:
+
+.. code-block:: python
+   from compression import zstd
```
```suggestion

.. code-block:: python

   from compression import zstd
```

StanFromIreland commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,772 @@
+Compressing data in memory:
+
+.. code-block:: python
+   from compression import zstd
```
```suggestion

Compressing data in memory:

.. code-block:: python

   from compression import zstd
```

StanFromIreland commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,772 @@
+Incremental compression:
+
+.. code-block:: python
+   from compression import zstd
```
```suggestion

.. code-block:: python

   from compression import zstd
```

StanFromIreland commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,772 @@
+Writing compressed data to an already-open file:
+
+.. code-block:: python
+   from compression import zstd
```
```suggestion

.. code-block:: python

   from compression import zstd
```

StanFromIreland commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,772 @@
+Creating a compressed file using compression parameters:
+
+.. code-block:: python
+   from compression import zstd
```
```suggestion

.. code-block:: python

   from compression import zstd
```

StanFromIreland commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,772 @@
+      Log size of each bucket in the long distance matcher hash table for
+      collision resolution. Larger values improve collision resolution but
+      decrease compression speed.
+   .. attribute:: ldm_hash_rate_log
```
```suggestion
      decrease compression speed.

   .. attribute:: ldm_hash_rate_log
```

StanFromIreland commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,772 @@
+      The impact of this field depends on the selected :class:`Strategy`.
+
+      For strategies :attr:`~Strategy.btopt`, :attr:`~Strategy.btultra` and
+      :attr:`~Strategy.btultra2`, the values is the length of a match
```
```suggestion
      :attr:`~Strategy.btultra2`, the value is the length of a match
```
Grammar

StanFromIreland commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,772 @@
+   .. attribute:: overlap_log
+
+      Sets how much data is reloaded from previous jobs (threads) for new jobs
+      to be used by the look behind window during compression. This values is
```
```suggestion
      to be used by the look behind window during compression. This value is
```
Grammar

StanFromIreland commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,772 @@
+      Select how many threads will be spawned to compress in parallel. When
+      :attr:`~.nb_workers` >= 1, enables multi-threaded compression, 1
+      means "1-thread multi-threaded mode". More workers improve speed, but
+      also increases memory usage and slightly reduce compression ratio.
```
```suggestion
      also increase memory usage and slightly reduce compression ratio.
```
Grammar

AA-Turner replied:
there should be agreement between memory usage and compression ratio, either increases/reduces or increase/reduce[ the].

StanFromIreland commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,772 @@
+   A :class:`ZstdFile` can wrap an already-open :term:`file object`, or operate
+   directly on a named file. The *file* argument specifies either the file
+   object to wrap, or the name of the file to open (as a :class:`str`,
+   :class:`bytes` or :term:`path-like <path-like object>` object). When
```
```suggestion
   :class:`bytes` or :term:`path-like <path-like object>` object). If
```
Simpler IMO

StanFromIreland reviewed (commented):
Some grammar notes.

Code block directives need a line after.

Please settle on gaps between descriptions for the file, currently it seems random, some parts use two and others one.

emmatyping committed:
Apply suggestions from reviewers

Co-authored-by: Adam Turner <9087854+AA-Turner@users.noreply.github.com>
Co-authored-by: Stan Ulbrych <89152624+StanFromIreland@users.noreply.github.com>

StanFromIreland commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,780 @@
+
+--------------
+
+This module provides classes and convenience functions for compressing and
```
```suggestion
This module provides classes and functions for compressing and
```
Unnecessary and awkward

StanFromIreland commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,780 @@
+This module provides classes and convenience functions for compressing and
+decompressing data using the Zstandard (or "zstd") compression algorithm. Also
+included is a file interface supporting reading and writing contents of ``.zst``
+files created from the :program:`zstd` utility, as well as raw zstd compressed
```
```suggestion
files created with the :program:`zstd` utility, as well as raw zstd compressed
```
```suggestion
files created by the :program:`zstd` utility, as well as raw zstd compressed
```
Grammar

emmatyping replied:
TIL you can provide multiple suggestions in a single comment, that's neat!

StanFromIreland commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,780 @@
+--------------
+
+This module provides classes and convenience functions for compressing and
+decompressing data using the Zstandard (or "zstd") compression algorithm. Also
```
```suggestion
decompressing data using the Zstandard (or *zstd*) compression algorithm. Also
```
Italics are more fitting here

StanFromIreland commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,780 @@
+
+This module provides classes and convenience functions for compressing and
+decompressing data using the Zstandard (or "zstd") compression algorithm. Also
+included is a file interface supporting reading and writing contents of ``.zst``
```
```suggestion
included is a file interface that supports reading and writing the contents of ``.zst`` files.
```

StanFromIreland commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,780 @@
+  :class:`Strategy` classes for setting advanced (de)compression parameters.
+
+
+.. exception:: ZstdError
```
This should get a little introduction like other modules.

"This module provides the following exception: "

Or something along those lines.

emmatyping replied:
I looked at a few random modules and they didn't really introduce exceptions. I'll add a section however to match the rest of the document.

StanFromIreland commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,780 @@
+   The mode argument can be either 'r' for reading (default), 'w' for
+   overwriting, 'a' for appending, or 'x' for exclusive creation. These can
+   equivalently be given as 'rb', 'wb', 'ab', and 'xb' respectively. You may
+   also open in text mode with 'rt', 'wt', 'at', and 'xt' respectively.
```
```suggestion
   The mode argument can be either ``'r'`` for reading (default), ``'w'`` for
   overwriting, 'a' for appending, or ``'x'`` for exclusive creation. These can
   equivalently be given as ``'rb'``, ``'wb'``, ``'ab'``, and ``'xb'`` respectively. You may
   also open in text mode with ``'rt'``, ``'wt'``, ``'at'``, and ``'xt'`` respectively.
```
As they are python strings

StanFromIreland commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,780 @@
+   also open in text mode with 'rt', 'wt', 'at', and 'xt' respectively.
+
+   When opening a file for reading, the *options* argument can be a dictionary
+   providing advanced decompression parameters, see
```
```suggestion
   providing advanced decompression parameters; see
```

StanFromIreland commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,780 @@
+   argument should not be used.
+
+   When opening a file for writing, the *options* argument can be a dictionary
+   providing advanced decompression parameters, see
```
```suggestion
   providing advanced decompression parameters; see
```

StanFromIreland commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,780 @@
+   *zstd_dict* argument is a :class:`ZstdDict` instance to be used during
+   compression.
+
+   For binary mode, this function is equivalent to the :class:`ZstdFile`
```
```suggestion
   In binary mode, this function is equivalent to the :class:`ZstdFile`
```

StanFromIreland commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,780 @@
+   constructor: ``ZstdFile(file, mode, ...)``. In this case, the
+   *encoding*, *errors* and *newline* parameters must not be provided.
+
+   For text mode, a :class:`ZstdFile` object is created, and wrapped in an
```
```suggestion
   In text mode, a :class:`ZstdFile` object is created, and wrapped in an
```

StanFromIreland commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,780 @@
+
+   For text mode, a :class:`ZstdFile` object is created, and wrapped in an
+   :class:`io.TextIOWrapper` instance with the specified encoding, error handling
+   behavior, and line ending(s).
```
```suggestion
   behavior, and line endings.
```
Unnecessary IMO

StanFromIreland commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,780 @@
+
+   For binary mode, this function is equivalent to the :class:`ZstdFile`
+   constructor: ``ZstdFile(file, mode, ...)``. In this case, the
+   *encoding*, *errors* and *newline* parameters must not be provided.
```
```suggestion
   *encoding*, *errors*, and *newline* parameters must not be provided.
```

emmatyping committed:
Apply suggestions from reviewers

Co-authored-by: Stan Ulbrych <89152624+StanFromIreland@users.noreply.github.com>

StanFromIreland commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,780 @@
+This module provides classes and functions for compressing and
+decompressing data using the Zstandard (or *zstd*) compression algorithm. Also
+included is a file interface that supports reading and writing the contents of ``.zst`` files.
+files created by the :program:`zstd` utility, as well as raw zstd compressed
```
```suggestion
included is a file interface that supports reading and writing the contents of ``.zst`` files,
files created by the :program:`zstd` utility, as well as raw zstd compressed
```
I may have introduced that one in a suggestion, this also needs to be wrapped I presume. (It would be handy if github provided a ruler in their diff viewer.)

mwichmann replied:
"files, files" doesn't seem a comfortable sequence.

emmatyping replied:
I think there was a misunderstanding in suggested reviews that mangled this but I've pushed a fix for this.

mwichmann commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,780 @@
+
+.. module:: compression.zstd
+   :synopsis: Low level interface to compression and decompression routines in
+              Meta's zstd library
```
Wouldn't call out Meta here. The reference manual for the zstd itself doesn't mention a company name.  On the other hand, it might not be a bad idea to reference that manual somewhere (https://facebook.github.io/zstd/zstd_manual.html).

hugovk replied:
```suggestion
              the Zstandard library
```

emmatyping replied:
Hm, the *library* is called `zstd`, perhaps we should say "in the zstd library"?

brainwane replied:
I second the suggestion to link to the [zstd manual](https://facebook.github.io/zstd/zstd_manual.html), and to copy or quote the short explanation: "... is a fast lossless compression algorithm, targeting real-time compression scenarios at zlib-level and better compression ratios."

emmatyping replied:
I ended up opening https://github.com/facebook/zstd/issues/4388 because it turns out the SEO first result is not the latest version linked from https://facebook.github.io/zstd/, which is version specific.

For now I think I'll leave out a link to the manual and add one later once that issue is resolved.

I've removed the Meta reference and said "in the zstd library."

emmatyping replied:
Yann was quick! That issue is resolved and I just pushed up the requested changes.

mwichmann commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,780 @@
+   to read from or write to.
+
+   The mode argument can be either ``'r'`` for reading (default), ``'w'`` for
+   overwriting, 'a' for appending, or ``'x'`` for exclusive creation. These can
```
markup on append mode is inconsistent

hugovk replied:
```suggestion
   overwriting, ``'a'`` for appending, or ``'x'`` for exclusive creation. These can
```

mwichmann commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,780 @@
+   :class:`DecompressionParameter` for detailed information about supported
+   parameters. The *zstd_dict* argument is a :class:`ZstdDict` instance to be
+   used during decompression. When opening a file for reading, the *level*
+   argument should not be used.
```
This may end up being an implementation question: if it "should not be used", does that mean it's ignored for reading (if so, perhaps say so)? Or does it cause problems if specified?

emmatyping replied:
Yeah, I think it would be good to say that it will raise an exception if passed during decompression.

emmatyping replied:
I re-worded this a little to call out that a TypeError is raised.

mwichmann commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,780 @@
+   When opening a file for reading, the *options* argument can be a dictionary
+   providing advanced decompression parameters, see
+   :class:`DecompressionParameter` for detailed information about supported
+   parameters. The *zstd_dict* argument is a :class:`ZstdDict` instance to be
```
Couldn't lines 59-62, 66-69, 102-105 and 109-112 be combined into a single clause, leaving only the specifics for reading and writing as unique? At the very least, four of these clauses in a short space seems a lot, two may be okay.

emmatyping replied:
Sorry I think the line numbers have changed and Github does not do a good job at showing enough context for your comment, what were you proposing?

mwichmann replied:
Yeah, line #s are always unsafe to add to comments, sorry.  It just sounded quite repetitive to, four times, say "When opening a file for... " , followed by the stanza about passing a dict of parameters. It doesn't harm anything, and if there's no obvious way to factor out some of that, then nevermind (I know they are all different combinations:  read/text/decompress, read/bytes/decompress, write/text/compress, write/bytes/compress)

mwichmann commented on Doc/library/compression.rst:
```diff
@@ -0,0 +1,18 @@
+
+.. attention::
+
+   The :mod:`!compression` package is the new location for the data compression
```
"New" always seems such a temporal term.  I'm just spitballing here, but how about something like: "The compression package provides a unified interface to several data compression routines. Some of these have historically been available as separate modules; those will continue to be available under their original names for compatibility reasons, and will not be removed without a deprecation cycle".

emmatyping replied:
Hm, I definitely understand "new" may not be the right word here. I worry if we say "a unified interface to ... compression routines" that people will expect `compression` to hold some high level `compress` function that abstracts over the various implementations, which is not the case.

Perhaps
"The compression package contains the canonical compression modules containing interfaces to several different algorithms. Some of these have historically been available as separate modules; those will continue to be available under their original names for compatibility reasons, and will not be removed without a deprecation cycle. The compression.* import names are encouraged for use where practical."

emmatyping replied:
I pushed up a slightly tweaked version of my suggestion. Please let me know what you think!

mwichmann replied:
the intro looks good

mwichmann commented on Doc/library/compression.rst:
```diff
@@ -0,0 +1,18 @@
+   The :mod:`!compression` package is the new location for the data compression
+   modules in the standard library, listed below. The existing modules are not
+   deprecated and will not be removed before Python 3.19. The new ``compression.*``
+   import names are encouraged for use where practicable.
```
Just "practical".

StanFromIreland replied:
*practicable* is grammatically correct, no? I think it’s fine.

hugovk replied:
Both are grammatical with slightly different meanings. "Practical" is more direct clearer.

Practicable: "are encouraged for use where [it could generally be done]"

Practical: "are encouraged for use where [it can actually be done]"

```suggestion
   import names are encouraged for use where practical.
```

AA-Turner replied:
I suggested practicable as I think it covers the intent more closely, compare *Able to be done or put into practice successfully; feasible* (“Practicable, Adj., Sense 1.”) with *Of, relating to practice or action, as opposed to speculation or theory.* (“Practical, Adj., Sense I.1.a.”), both copyright the Oxford English Dictionary. If both Hugo and Mats prefer changing, though, I have no strong objection.

mwichmann replied:
I certainly won't raise objections if it stays. I've spent years trying to convince myself to use the "simpler" word in docs when there are a couple of choices, is all it was.

hugovk commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,780 @@
+=============================================================================
+
+.. module:: compression.zstd
+   :synopsis: Low level interface to compression and decompression routines in
```
```suggestion
   :synopsis: Low-level interface to compression and decompression routines in
```

hugovk commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,780 @@
+   respectively.
+
+   If *file* is a file object (rather than an actual file name), a mode of
+   ``"w"`` does not truncate the file, and is instead equivalent to ``"a"``.
```
Single quotes were used for `open`, be consistent one way or the other.

hugovk commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,780 @@
+
+      .. note:: While calling :meth:`peek` does not change the file position of
+         the :class:`ZstdFile`, it may change the position of the underlying
+         file object (e.g. if the :class:`ZstdFile` was constructed by passing a
```
Avoid Latin abbreviations: https://devguide.python.org/documentation/style-guide/#use-simple-language
```suggestion
         file object (for example, if the :class:`ZstdFile` was constructed by passing a
```

hugovk commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,780 @@
+   meaning of *dict_content*. ``True`` means *dict_content* is a "raw content"
+   dictionary, without any format restrictions. ``False`` means *dict_content*
+   is an ordinary Zstandard dictionary, created from Zstandard functions, e.g.
+   :func:`train_dict` or the ``zstd`` CLI.
```
```suggestion
   is an ordinary Zstandard dictionary, created from Zstandard functions,
   for example, :func:`train_dict` or the ``zstd`` CLI.
```

hugovk commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,780 @@
+
+        Digesting a dictionary is a costly operation. These two attributes can
+        control how the dictionary is loaded to the compressor, by passing them
+        as the ``zstd_dict`` argument, e.g.
```
```suggestion
        as the ``zstd_dict`` argument, for example,
```

hugovk commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,780 @@
+   data as a :class:`bytes` object.
+
+   The *level* argument is an int object controlling the level of
+   compression. Please refer to :meth:`CompressionParameter.bounds` to get the
```
Usually no need for "please" in technical documentation:
```suggestion
   compression. Refer to :meth:`CompressionParameter.bounds` to get the
```

* https://developers.google.com/style/word-list#please
* https://learn.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/p/please
* https://support.apple.com/en-gb/guide/applestyleguide/apsg4473eab0/web
* https://stylepedia.net/style/

hugovk commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,780 @@
+   module-level function :func:`compress`.
+
+   The *level* argument is an int object controlling the level of
+   compression. Please refer to :meth:`CompressionParameter.bounds` to get the
```
```suggestion
   compression. Refer to :meth:`CompressionParameter.bounds` to get the
```

hugovk commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,780 @@
+   generating the Zstandard dictionary.
+
+   The *dict_size* argument, an integer, is the maximum size (in bytes) the
+   Zstandard dictionary should be. Please see :func:`train_dict` for
```
```suggestion
   Zstandard dictionary should be. See :func:`train_dict` for
```

hugovk commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,780 @@
+      The returned data should be concatenated with the output of any previous
+      calls to :meth:`~.decompress`.
+
+      If *max_length* is nonnegative, returns at most *max_length*
```
```suggestion
      If *max_length* is non-negative, returns at most *max_length*
```

hugovk commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,780 @@
+
+
+Zstandard Dictionaries
+----------------------
```
```suggestion
Zstandard dictionaries
----------------------
```
https://devguide.python.org/documentation/style-guide/#capitalization

emmatyping replied:
Thanks, missed this!

hugovk commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,780 @@
+
+   The *dict_size* argument, an integer, is the maximum size (in bytes) the
+   Zstandard dictionary should be. The Zstandard documentation suggests an
+   absolute maximum of no more than 100KB, but the maximum can often be smaller
```
```suggestion
   absolute maximum of no more than 100 KB, but the maximum can often be smaller
```

hugovk commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,780 @@
+
+    .. attribute:: dict_id
+
+        Identifier of the Zstandard dictionary, a int value between 0 and .
```
Incomplete sentence:
```suggestion
        Identifier of the Zstandard dictionary, an int value between zero and .
```

emmatyping replied:
Hm, there was another incomplete sentence above too. I think these might have been mangled in a review suggestion (or perhaps in a change I made?), I'll take a look at past commits.

hugovk commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,780 @@
+        ``compress(data, zstd_dict=zd.as_digested_dict)``.
+
+        If don't use one of these attributes, an **undigested** dictionary is
+        passed by default.
```
"If don't use" -- some missing words here? Or reword the sentence?

emmatyping replied:
I re-wrote this section because it could use some attention.

hugovk commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,780 @@
+   .. attribute:: window_log
+
+      Maximum allowed back-reference distance the compressor can use when
+      compressing data, expressed as power of 2, ``1 << window_log`` bytes. This
```
```suggestion
      compressing data, expressed as power of two, ``1 << window_log`` bytes. This
```

hugovk commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,780 @@
+
+   .. attribute:: hash_log
+
+      Size of the initial probe table, as a power of 2. The resulting memory
```
```suggestion
      Size of the initial probe table, as a power of two. The resulting memory
```

hugovk commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,780 @@
+
+   .. attribute:: chain_log
+
+      Size of the multi-probe search table, as a power of 2. The resulting
```
```suggestion
      Size of the multi-probe search table, as a power of two. The resulting
```

hugovk commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,780 @@
+
+   .. attribute:: search_log
+
+      Number of search attempts, as a power of 2. More attempts result in
```
```suggestion
      Number of search attempts, as a power of two. More attempts result in
```

hugovk commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,780 @@
+   The :meth:`~.bounds` method can be used on any attribute to get the valid
+   values for that parameter.
+
+   Setting any parameter to "0" causes zstd to dynamically select a value
```
```suggestion
   Setting any parameter to zero causes zstd to dynamically select a value
```

hugovk commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,780 @@
+   .. attribute:: compression_level
+
+      A high-level means of setting other compression parameters that affect
+      the speed and ratio of compressing data. Setting the level to 0 uses the
```
```suggestion
      the speed and ratio of compressing data. Setting the level to zero uses the
```

hugovk commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,780 @@
+      find matches of smaller size, it just tweaks its search algorithm to look
+      for this size and larger. Note that currently, for all strategies
+      < :attr:`~Strategy.btopt`, the effective minimum is ``4``, for all
+      strategies > :attr:`~Strategy.fast`, the effective maximum is ``6``.
```
Two "Note that" here, can we remove or reword one?

emmatyping replied:
Yeah I was going to just delete the second "Note that currently"

hugovk commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,780 @@
+
+   .. attribute:: ldm_hash_log
+
+      Size of the table for long distance matching, as a power of 2. Larger
```
```suggestion
      Size of the table for long distance matching, as a power of two. Larger
```

hugovk commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,780 @@
+      :attr:`~CompressionParameter.window_log` to 128 MiB except when expressly
+      set to a different value. This setting is enabled by default if
+      :attr:`~CompressionParameter.window_log` >= 128 MiB and the compression
+      strategy >= :attr:`~Strategy.btopt` (compression level 16+).
```
Don't link the second one in the same paragraph: 
```suggestion
      Enabling this parameter increases default
      :attr:`~CompressionParameter.window_log` to 128 MiB except when expressly
      set to a different value. This setting is enabled by default if
      :attr:`!window_log` >= 128 MiB and the compression
      strategy >= :attr:`~Strategy.btopt` (compression level 16+).
```
https://devguide.python.org/documentation/style-guide/#links

hugovk commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,780 @@
+
+      Uncompressed content size will be written into frame header whenever
+      known. This flag currently has no effect.
+
```
If this has no effect, I'm wondering if we need to document this. Do we even need it in the API?

emmatyping replied:
I think I would like to make this take effect and update this in a future PR. Happy to remove from the docs for now.

StanFromIreland replied:
That would be best IMO.

hugovk commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,780 @@
+
+   .. attribute:: checksum_flag
+
+      A 4-byte checksum using XXHash64 of the uncompressed content is written
```
```suggestion
      A four-byte checksum using XXHash64 of the uncompressed content is written
```

hugovk commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,780 @@
+   .. attribute:: nb_workers
+
+      Select how many threads will be spawned to compress in parallel. When
+      :attr:`~.nb_workers` >= 1, enables multi-threaded compression, 1
```
Don't link to self:
```suggestion
      :attr:`!nb_workers` >= 1, enables multi-threaded compression, 1
```

hugovk commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,780 @@
+
+.. code-block:: python
+
+   from compression import zstd
```
Perhaps a newline after imports for readability?
```suggestion
   from compression import zstd

   with zstd.open("file.zst") as f:
       file_content = f.read()

Creating a compressed file:

.. code-block:: python

   from compression import zstd

   data = b"Insert Data Here"
   with zstd.open("file.zst", "w") as f:
       f.write(data)

Compressing data in memory:

.. code-block:: python

   from compression import zstd

   data_in = b"Insert Data Here"
   data_out = zstd.compress(data_in)

Incremental compression:

.. code-block:: python

   from compression import zstd

   comp = zstd.ZstdCompressor()
   out1 = comp.compress(b"Some data\n")
   out2 = comp.compress(b"Another piece of data\n")
   out3 = comp.compress(b"Even more data\n")
   out4 = comp.flush()
   # Concatenate all the partial results:
   result = b"".join([out1, out2, out3, out4])

Writing compressed data to an already-open file:

.. code-block:: python

   from compression import zstd

   with open("file.zst", "wb") as f:
       f.write(b"This data will not be compressed\n")
       with zstd.open(f, "w") as zstf:
           zstf.write(b"This *will* be compressed\n")
       f.write(b"Not compressed\n")

Creating a compressed file using compression parameters:

.. code-block:: python

   from compression import zstd

```

hugovk commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,780 @@
+files created by the :program:`zstd` utility, as well as raw zstd compressed
+streams.
+
+The :mod:`compression.zstd` module contains:
```
No need to self-link:
```suggestion
The :mod:`!compression.zstd` module contains:
```

hugovk commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,780 @@
+   When opening a file for reading, the *options* argument can be a dictionary
+   providing advanced decompression parameters, see
+   :class:`DecompressionParameter` for detailed information about supported
+   parameters. The *zstd_dict* argument is a :class:`ZstdDict` instance to be
```
```suggestion
   parameters. The *zstd_dict* argument is a :class:`!ZstdDict` instance to be
```

hugovk commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,780 @@
+   :class:`CompressionParameter` for detailed information about supported
+   parameters. The *level* argument is the compression level to use when
+   writing compressed data. Only one of *level* or *options* may be passed. The
+   *zstd_dict* argument is a :class:`ZstdDict` instance to be used during
```
```suggestion
   *zstd_dict* argument is a :class:`!ZstdDict` instance to be used during
```

hugovk commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,780 @@
+   When opening a file for writing, the *options*, *zstd_dict* and *level*
+   arguments have the same meanings as for :class:`ZstdCompressor`.
+
+   :class:`ZstdFile` supports all the members specified by
```
```suggestion
   :class:`!ZstdFile` supports all the members specified by
```

hugovk commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,780 @@
+      Compress *data* (a :term:`bytes-like object`), returning a :class:`bytes`
+      object if possible, or an empty byte string otherwise. Some of *data* may
+      be buffered internally, for use in later calls to
+      :meth:`~.compress` and :meth:`~.flush`. The
```
```suggestion
      :meth:`!compress` and :meth:`~.flush`. The
```

hugovk commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,780 @@
+      compressed frames, unlike the :func:`decompress` function and
+      :class:`ZstdFile` class. To decompress a multi-frame input, you should
+      use :func:`decompress`, :class:`ZstdFile` if working with a
+      :term:`file object`, or multiple :class:`ZstdDecompressor` instances.
```
```suggestion
      :term:`file object`, or multiple :class:`!ZstdDecompressor` instances.
```

hugovk commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,780 @@
+      uncompressed data as bytes. Some of *data* may be buffered
+      internally, for use in later calls to :meth:`~.decompress`.
+      The returned data should be concatenated with the output of any previous
+      calls to :meth:`~.decompress`.
```
```suggestion
      internally, for use in later calls to :meth:`!decompress`.
      The returned data should be concatenated with the output of any previous
      calls to :meth:`!decompress`.
```

hugovk commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,780 @@
+
+        .. note::
+
+            The meaning of ``0`` for :attr:`ZstdDict.dict_id` is different from
```
```suggestion
            The meaning of ``0`` for :attr:`!ZstdDict.dict_id` is different from
```

hugovk commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,780 @@
+      Minimum size of searched matches. Larger values increase compression and
+      decompression speed, but decrease ratio. Note that Zstandard can still
+      find matches of smaller size, it just tweaks its search algorithm to look
+      for this size and larger. Note that currently, for all strategies
```
Everything in these (versioned) docs should be "currently", no need to say so unless it's changing.

```suggestion
      for this size and larger. Note that for all strategies
```

hugovk commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,780 @@
+.. class:: FrameInfo
+
+   Metadata related to a Zstandard frame. There are currently two attributes
+   containing metadata related to Zstandard frames.
```
```suggestion
   Metadata related to a Zstandard frame. There are two attributes
   containing metadata related to Zstandard frames.
```
Fine to leave a count, but these are often things forgotten when adding a new one :)
```suggestion
   Metadata related to a Zstandard frame. Here are the attributes
   containing metadata related to Zstandard frames.
```
Or perhaps remove altogether:
```suggestion
   Metadata related to a Zstandard frame.
```

hugovk commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,780 @@
+
+.. attribute:: COMPRESSION_LEVEL_DEFAULT
+
+   The default compression level for Zstandard, currently '3'.
```
```suggestion
   The default compression level for Zstandard: '3'.
```

StanFromIreland replied:
```suggestion
   The default compression level for Zstandard: ``'3'``.
```
Since it is the return value, unless it is an integer than this is much better:
```suggestion
   The default compression level for Zstandard: ``3``.
```

emmatyping committed:
Apply suggestions from reviewers

Co-authored-by: Hugo van Kemenade <1324225+hugovk@users.noreply.github.com>
Co-authored-by: Stan Ulbrych <89152624+StanFromIreland@users.noreply.github.com>

emmatyping committed:
Don't reference self when referring to items

Co-authored-by: Hugo van Kemenade <1324225+hugovk@users.noreply.github.com>

emmatyping committed:
Updates to respond to review

- Rewrite the digested/undigested dict section
- Add a section for exceptions
- wrap lines to ~80 chars
- clarify that an exception is raised if level is passed to a decompressor
- make quotes in docs for open vs ZstdFile consistent
- Remove currently and repeated "note"

vadmium commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,778 @@
+   in which case the named file is opened, or it can be an existing file object
+   to read from or write to.
+
+   The mode argument can be either ``'r'`` for reading (default), ``'w'`` for
```
Minor discordance with 'rb' default in signature. My preference would be to make 'rb', 'wb', etc the preferred form (explicit _b_ for binary), and then list 'r' etc as secondary equivalents. The other option is to change the signature to 'r'.

vadmium commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,778 @@
+      parameter. This method should be called on the attribute you wish to
+      retrieve the bounds of. For example, to get the valid values for
+      :attr:`~.window_log_max`, one may check the result of
+      ``CompressionParameter.window_log_max.bounds()``.
```
```suggestion
      ``DecompressionParameter.window_log_max.bounds()``.
```

emmatyping replied:
Agh! Thanks for catching this.

StanFromIreland commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,778 @@
+   documented as part of the :class:`CompressionParameter` documentation.
+
+   The *zstd_dict* argument is an instance of :class:`ZstdDict`, a Zstandard
+   dictionary, containing trained data to improve compression efficiency. The
```
```suggestion
   The *zstd_dict* argument is an instance of :class:`ZstdDict`
   containing trained data to improve compression efficiency. The
```
We don't need that, you can click the link for an explanation.

StanFromIreland commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,778 @@
+   documented as part of the :class:`CompressionParameter` documentation.
+
+   The *zstd_dict* argument is an instance of :class:`ZstdDict`, a Zstandard
+   dictionary, containing trained data to improve compression efficiency. The
```
```suggestion
   The *zstd_dict* argument is an instance of :class:`ZstdDict`
   containing trained data to improve compression efficiency. The
```
ditto

StanFromIreland commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,778 @@
+   documentation.
+
+   The *zstd_dict* argument is an instance of :class:`ZstdDict`, a Zstandard
+   dictionary, containing trained data used during compression. This must be
```
```suggestion
   The *zstd_dict* argument is an instance of :class:`ZstdDict`
   containing trained data used during compression. This must be
```
ditto

StanFromIreland commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,778 @@
+   documentation.
+
+   The *zstd_dict* argument is an instance of :class:`ZstdDict`, a Zstandard
+   dictionary, containing trained data used during compression. This must be
```
```suggestion
   The *zstd_dict* argument is an instance of :class:`ZstdDict`
   containing trained data used during compression. This must be
```
yet again

vadmium commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,778 @@
+
+      A high-level means of setting other compression parameters that affect
+      the speed and ratio of compressing data. Setting the level to zero uses
+      the default :attr:`COMPRESSION_LEVEL_DEFAULT`.
```
```suggestion
      the speed and ratio of compressing data. Setting the level to zero uses
      :attr:`COMPRESSION_LEVEL_DEFAULT`.
```

vadmium commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,778 @@
+   The :meth:`~.bounds` method can be used on any attribute to get the valid
+   values for that parameter.
+
+   Setting any parameter to zero causes zstd to dynamically select a value
```
Assuming _zstd_ refers to something in the module implementation, this leads to the question of whether omitting parameters is supported. I think the documentation should clarify this. The one example at the bottom of the page that does pass an _options_ dictionary does not set a value for every parameter.

vadmium commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,778 @@
+           ``ldm_min_match``, ``ldm_bucket_size_log``, ``ldm_hash_rate_log``,
+           and some non-public parameters.
+         - None
+       * - :class:`!ZstdDict` internally caches the dictionary
```
```suggestion
       * - :class:`!ZstdDict` internally caches the dictionary?
```

emmatyping replied:
Sorry, could you explain the question mark here?

vadmium commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,778 @@
+   dictionary, containing trained data to improve compression efficiency. The
+   function :func:`train_dict` can be used to generate a Zstandard dictionary.
+
+   .. attribute:: CONTINUE
```
I feel like these three mode attributes could be listed less abruptly, to sound more like object attributes and less like methods. Perhaps just move them after the two methods that they would be used with? Or describe them as flush modes.

emmatyping replied:
Yeah I think it would be good to move them after methods.

vadmium commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,778 @@
+   .. method:: compress(data, mode=ZstdCompressor.CONTINUE)
+
+      Compress *data* (a :term:`bytes-like object`), returning a :class:`bytes`
+      object if possible, or an empty byte string otherwise. Some of *data* may
```
I presume the empty byte string is still a _bytes_ object, no? Otherwise this looks like a mistake.

```suggestion
      Compress *data* (a :term:`bytes-like object`), returning a :class:`bytes`
      object. Some of *data* may
```

emmatyping replied:
Yes, but it is important to include that it may return an empty bytes object (some people may expect it to always make forward progress)

emmatyping replied:
I pushed up something that should hopefully be a bit clearer?

emmatyping committed:
Remove outdated paragraph

vadmium commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,775 @@
+   files), Zstandard dictionaries can improve compression ratios and speed
+   significantly.
+
+   The *samples* argument (an iterable of :class:`bytes`), is the population of
```
```suggestion
   The *samples* argument (an iterable of :class:`bytes` objects), is the population of
```

Or did you actually mean things like a single _bytes_ object is accepted, which is iterable of byte values? If not, is there any difference between iterating multiple _bytes_ strings versus iterating a single concatenated string?

Similarly under _finalize_dict_.

vadmium commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,775 @@
+   meaning of *dict_content*. ``True`` means *dict_content* is a "raw content"
+   dictionary, without any format restrictions. ``False`` means *dict_content*
+   is an ordinary Zstandard dictionary, created from Zstandard functions,
+   for example, :func:`train_dict` or the ``zstd`` CLI.
```
Presume _zstd_ is a program external to Python, or is there a _compression.zstd_ module CLI? Need better contrast with _train_dict_ which I presume _is_ the function in the Python module.

vadmium commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,775 @@
+
+    .. attribute:: dict_id
+
+        Identifier of the Zstandard dictionary, an int value between zero and .
```
```suggestion
        Identifier of the Zstandard dictionary, a non-negative int value.
```

vadmium commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,775 @@
+
+    .. attribute:: as_digested_dict
+
+        Load as a digested dictionary, see below.
```
```suggestion
        Load as a digested dictionary.
```

I don’t see what you might be referring to, unless you mean the text about digestion above the list of attributes.

vadmium commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,775 @@
+      When you have finished providing data to the compressor, call the
+      :meth:`~.flush` method to finish the compression process.
+
+   .. method:: flush(mode)
```
Is _mode_ optional? See incremental compression example at the bottom.

emmatyping replied:
Yes, I have made it optional, good catch!

vadmium commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,775 @@
+   If *file* is a file object (rather than an actual file name), a mode of
+   ``'w'`` does not truncate the file, and is instead equivalent to ``'a'``.
+
+   When opening a file for reading, the *options* argument can be a dictionary
```
```suggestion
   When decompressing, the *options* argument can be a dictionary
```

I presume this also applies when wrapping an existing file object, not just when opening a new one. Likewise for next paragraph on writing/compressing.

emmatyping replied:
I think I would like to keep it termed as reading since this is a high-level API. Perhaps we can say "When reading from a file like object"?

emmatyping replied:
I just shortened to "when reading/writing"

vadmium commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,775 @@
+
+   When opening a file for reading, the *options* argument can be a dictionary
+   providing advanced decompression parameters, see
+   :class:`DecompressionParameter` for detailed information about supported
```
It took me a while (reading about ZstdCompressor) before I could guess how the _options_ dictionary might relate to (De)compressionParameter. I think you should explicitly say the keys are the enum members, like you have already hinted under Zstd(De)compressor. Or have a separate section explaining the common parameters like _options_ all in one place.

emmatyping replied:
I added some examples to the parameter enums which should hopefully help here. There are a lot of places where the docs refer to them and I don't want to have to repeat that N times across the docs.

vadmium commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,775 @@
+   When opening a file for reading, the *options* argument can be a dictionary
+   providing advanced decompression parameters, see
+   :class:`DecompressionParameter` for detailed information about supported
+   parameters. The *zstd_dict* argument is a :class:`!ZstdDict` instance to be
```
I suspect @hugovk removed the link by mistake.

```suggestion
   parameters. The *zstd_dict* argument is a :class:`ZstdDict` instance to be
```

Also applies to the next paragraph for compress mode.

emmatyping committed:
Remove Zstandard dictionary after ZstdDict

Co-authored-by: Stan Ulbrych <89152624+StanFromIreland@users.noreply.github.com>

vadmium commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,775 @@
+      The returned data should be concatenated with the output of any previous
+      calls to :meth:`!decompress`.
+
+      If *max_length* is non-negative, returns at most *max_length*
```
```suggestion
      If *max_length* is non-negative, the method returns at most *max_length*
```

vadmium commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,775 @@
+   .. attribute:: needs_input
+
+      ``False`` if the :meth:`.decompress` method can provide more
+      decompressed data before requiring new uncompressed input.
```
```suggestion
      decompressed data before requiring new compressed input.
```

vadmium commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,775 @@
+       * -
+         - Digested dictionary
+         - Undigested dictionary
+       * - Advanced parameters of the compressor which may be overridden by
```
I wonder if these parameters are overridden in the first API call (when I presume the dictionary isn’t initially digested), or only on subsequent calls

emmatyping replied:
The dictionary is not automatically digested when compressing, and parameters are only overridden when compressing.

vadmium commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,775 @@
+
+   If passing a :class:`!ZstdDict` without any attribute, an undigested
+   dictionary is passed by default when compressing and a digested dictionary
+   is passed by default when decompressing.
```
Presume the digestion occurs when the dictionary is “loaded” in the first API call, so a digested dictionary can only be passed in the subsequent calls? This is not clear. Similar issue in description of _as_digested_dict_.

emmatyping committed:
Rewrite introduction to compression package to be more timeless

emmatyping committed:
Remove content_size_flag

emmatyping committed:
Merge branch 'zstd-docs' of github.com:emmatyping/cpython into zstd-docs

StanFromIreland commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,770 @@
+
+The :mod:`!compression.zstd` module contains:
+
+* The :func:`.open` function and :class:`ZstdFile` class for reading and
```
```suggestion
* The :func:`open` function and :class:`ZstdFile` class for reading and
```
It seems wrong yet it renders ok?

emmatyping replied:
I believe this is correct, ref https://www.sphinx-doc.org/en/master/usage/domains/python.html#target-resolution

StanFromIreland commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,770 @@
+   Open a Zstandard-compressed file in binary or text mode, returning a
+   :term:`file object`.
+
+   The *file* argument can be either an actual file name (given as a
```
```suggestion
   The *file* argument can be either a file name (given as a
```
[That *file name* is definitely an *actual file name*.](https://www.youtube.com/watch?v=vZw35VUBdzo)

emmatyping replied:
Thanks I missed updating this when I updated filename to file.

StanFromIreland commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,770 @@
+      :attr:`~.CONTINUE`, :attr:`~.FLUSH_BLOCK`,
+      or :attr:`~.FLUSH_FRAME`.
+
+      When you have finished providing data to the compressor, call the
```
```suggestion
      When all data has been provided to the compressor, call the
```
Formal tone

StanFromIreland commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,770 @@
+      zstd.CompressionParameter.checksum_flag: 1
+   }
+   with zstd.open("file.zst", "w", options=options) as f:
+       f.write(b"blah blah blah")
```
```suggestion
       f.write(b"Do your worst you worm!")
```
Why not make our example more fun when we have the chance, I like to come across these in the docs, the best I could think of was [this](https://www.youtube.com/watch?v=o13glRURgTE) ;-)

brainwane replied:
I also enjoy fun examples, and recommend something more friendly - maybe:

```suggestion
       f.write(b"Mind if I squeeze in?")
```

StanFromIreland reviewed (commented):
Looking pretty good now.

vadmium commented on Doc/library/compression.rst:
```diff
@@ -0,0 +1,18 @@
+these modules have historically been available as separate modules; those will
+continue to be available under their original names for compatibility reasons,
+and will not be removed without a deprecation cycle. The use of modules in
+:mod:`!compression` is encouraged where practical.
```
I think it would be better for the documentation of the four original modules to reflect their new preferred names. That way people reading e.g. the _gzip_ documentation will be aware of the new preferred module name.

emmatyping replied:
I think that is something we can add in the future, but not something I want to cover in this PR. I think the name change is an important piece of information people reading the compression module docs would want to know.

vadmium commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,770 @@
+.. class:: CompressionParameter()
+
+   An :class:`~enum.IntEnum` containing the advanced compression parameter
+   names that can be used when compressing data.
```
```suggestion
   keys that can be used when compressing data.
```

Presuming the members are numbers, it’s strange to call them “names”. Also applies under DecompressionParameter.

vadmium commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,770 @@
+      decompression speed, but decrease ratio. Note that Zstandard can still
+      find matches of smaller size, it just tweaks its search algorithm to look
+      for this size and larger. For all strategies < :attr:`~Strategy.btopt`,
+      the effective minimum is ``4``, for all strategies
```
```suggestion
      the effective minimum is ``4``; for all strategies
```

vadmium commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,770 @@
+      find matches of smaller size, it just tweaks its search algorithm to look
+      for this size and larger. For all strategies < :attr:`~Strategy.btopt`,
+      the effective minimum is ``4``, for all strategies
+      > :attr:`~Strategy.fast`, the effective maximum is ``6``.
```
Are the effective minimum and maximum limits on match sizes, or limits on the value of _min_match_?

emmatyping replied:
If you use a value outside that range then it will have no further effect. (i.e. if I am using Strategy.fast and set a min match of 7 it will have the same effect as 6).

vadmium commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,770 @@
+   .. note::
+
+      The values of attributes of :class:`!Strategy` are not necessarily stable
+      across zstd versions. Only the ordering of the attributes may be relied
```
Can we document this ordering and its direction then? I.e. which members are <= _dfast_ and > _dfast_, etc?

emmatyping replied:
They are listed as ordered, perhaps I should call that out?

emmatyping replied:
I now call that out, hopefully that is sufficient?

vadmium commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,770 @@
+      inputs by finding large matches at greater distances. It increases memory
+      usage and window size.
+
+      Enabling this parameter increases default
```
How do you enable and disable this, considering zero is supposed to have another special meaning?

emmatyping replied:
Ah, it doesn't here. I'll update the description above to call this out.

vadmium commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,770 @@
+
+   .. attribute:: window_log_max
+
+      The power of two maximum size of the window used during decompression.
```
I find this use of _power_ ambiguous. Should I set _window_log_max_ to 1024, since that is the 10th _power of two_? Or should I set it to 10, since two to the _power 10_ is 1024? Only after seeing _CompressionParameter.window_log_ and other parameters do I have enough confidence to guess it’s meant to be the exponent. The same problem exists with many of the __log_ CompressionParameter members too.

```suggestion
      The base-two logarithm of the maximum size of the window used during decompression.
```

Also, is there any detriment for giving a small value, e.g. slower operation, or perhaps absolute inability to decompress streams like is the case for _zlib_?

vadmium commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,770 @@
+      hash table. Larger values improve compression speed. Deviating far from
+      the default value will likely result in a compression ratio decrease.
+
+   .. attribute:: checksum_flag
```
Is this a flag you can turn on or off, and if so, how? The example at the bottom supplies a value of 1 without explanation. Similarly for _dict_id_flag_.

vadmium commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,770 @@
+
+      Select how many threads will be spawned to compress in parallel. When
+      :attr:`!nb_workers` >= 1, enables multi-threaded compression, 1
+      means "one-thread multi-threaded mode". More workers improve speed, but
```
Is “one-thread multi-threaded mode” just a degenerate case, or is it of some practical use? Either clarify what this means, or avoid drawing so much attention to it.

```suggestion
      :attr:`!nb_workers` >= 1, enables multi-threaded compression.
      The value 1 is unlikely to be useful. More workers improve speed, but
```

However if values >= 1 enable multi-threading, and 0 means dynamically select a value, how do I disable multi-threading?

emmatyping replied:
I think it is important to call attention to it as people may assume that it means do not start threads, when it does start a new thread.

serhiy-storchaka replied:
"nb_workers > 0" or "nb_workers is not zero" may be clearer than "nb_workers >= 1" . (Can it be negative?)

emmatyping replied:
I changed it to nb_workers > 0, great suggestion!

brainwane commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,770 @@
+   :class:`DecompressionParameter` for detailed information about supported
+   parameters. The *zstd_dict* argument is a :class:`ZstdDict` instance to be
+   used during decompression. When opening a file for reading, if the *level*
+   argument is passed a :exc:`!TypeError` will be raised.
```
```suggestion
   argument is passed, a :exc:`!TypeError` will be raised.
```

brainwane commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,770 @@
+
+   The *level* argument is an int object controlling the level of
+   compression. Refer to :meth:`CompressionParameter.bounds` to get the
+   values that can be passed for *level*. If advanced compression options are
```
Clarification: the upper and lower bounds are specific to that `CompressionParameter` instance? Is there a default range one is most likely to expect?

vadmium replied:
Reading between the lines, it sounds like _level_ may be another way of passing a _CompressionParameter.compression_level_ option, but it’s not clear. So I presume you’d actually have to call something like CompressionParameter.compression_level.bounds() for the _bounds_ method to work. Worth making explicit?

The only compression level values I’ve seen are 16+ (mentioned offhand under _enable_long_distance_matching_) and COMPRESSION_LEVEL_DEFAULT = 3. In order to use _level_ I think I would need more details, prior knowledge, or have to refer to external documentation or source code.

These comments also apply to the ZstdCompressor class below.

emmatyping replied:
> Worth making explicit?

Yeah, I'd like to do so but I don't think sphinx understands IntEnums so I can't refer to the bounds reference and say `CompressionParameter.compression_level.bounds()`. Perhaps I can say (paraphrasing) "use `CompressionParameter.compression_level.bounds()`; refer to `bounds()` (a reference link) for details."

vadmium replied:
Something like “_level_ is an alternative to _CompressionParameter.compression_level_” (if actually true) would be a start. You could also say “call _bounds_ on _compression_level_ to get the range of values for _level_” if you want, but that doesn’t seem so important if the link between _level_ and _compression_level_ is already make clear.

brainwane reviewed (commented):
I wasn't absolutely thorough in reviewing all the proposed text, but have a few suggestions for clarification. Thanks for the work.

emmatyping committed:
Apply suggestions from Sumana and Stan

Co-authored-by: Stan Ulbrych <89152624+StanFromIreland@users.noreply.github.com>
Co-authored-by: Sumana Harihareswara <sh@changeset.nyc>

vadmium commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,770 @@
+   :class:`DecompressionParameter` for detailed information about supported
+   parameters. The *zstd_dict* argument is a :class:`ZstdDict` instance to be
+   used during decompression. When opening a file for reading, if the *level*
+   argument is passed, a :exc:`!TypeError` will be raised.
```
```suggestion
   argument is not None, a :exc:`!TypeError` will be raised.
```

Assuming the default value is None as in the signature.

Similar confusion of omitting vs passing None for _level_ and _options_ in compress mode, and with the text encoding parameters in binary mode. And under ZstdFile and _compress_.

vadmium commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,770 @@
+      :attr:`~.CONTINUE`, :attr:`~.FLUSH_BLOCK`,
+      or :attr:`~.FLUSH_FRAME`.
+
+      When all data has been provided to the compressor, call the
```
Does this include when generating an empty frame, and _compress_ is never called? If so, maybe move this paragraph up to the general ZstdCompressor description, or down under the _flush_ method itself?

On the other hand, is calling _flush_ necessary if I have already passed FLUSH_FRAME to _compress_? Does that generate one or two frames?

emmatyping replied:
Yeah I think moving this up makes a lot of sense.

emmatyping replied:
I actually added a bit about what happens if you call compress with flush with both flushing a frame.

vadmium commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,770 @@
+   ``'w'`` does not truncate the file, and is instead equivalent to ``'a'``.
+
+   When opening a file for reading, the *options* argument can be a dictionary
+   providing advanced decompression parameters, see
```
```suggestion
   providing advanced decompression parameters; see
```

Also applies in next paragraph on compressing.

vadmium commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,770 @@
+
+   from compression import zstd
+
+   with open("file.zst", "wb") as f:
```
I suspect this may not be a valid “.zst” file, since it would have custom strings surrounding the Z Standard compressed data. Perhaps change the file name?

vadmium commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,770 @@
+        .. note::
+
+            The meaning of ``0`` for :attr:`!ZstdDict.dict_id` is different from
+            the ``dictionary_id`` argument to the :func:`get_frame_info`
```
There is only a _frame_buffer_ parameter listed for _get_frame_info_. Are you referring to the _FrameInfo.dictionary_id_ attribute of the object it presumably returns?

vadmium commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,770 @@
+      An int object representing the Zstandard dictionary ID needed for
+      decompressing the frame. ``0`` means the dictionary ID was not
+      recorded in the frame header, the frame may or may not need a dictionary
+      to be decoded, or the ID of such a dictionary is not specified.
```
```suggestion
      recorded in the frame header, or the frame does not need a dictionary
      to be decoded.
```

Most of these options seem the same as the first, unless I’m missing something.

vadmium commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,770 @@
+   parameters. The valid keys and values for compression parameters are
+   documented as part of the :class:`CompressionParameter` documentation.
+
+   The *zstd_dict* argument is an instance of :class:`ZstdDict`
```
```suggestion
   The *zstd_dict* argument is an optional instance of :class:`ZstdDict`
```

Also under _compress_, _decompress_, and ZstdDecompressor.

emmatyping committed:
Remove ref to Meta and clean up mode usage

emmatyping committed:
Apply suggestions from vadmium

Co-authored-by: Martin Panter <vadmium@users.noreply.github.com>

emmatyping committed:
Many updates to respond to review

- improve argument descriptions
- improve clarity about the meaning of parameter value 0s
- clarify behavior of compress/flush
- many more clarifications/wording improvements

emmatyping committed:
Add examples to (De)compressionParameter

emmatyping committed:
Add reference to zstd manual and blurb on algorithm

serhiy-storchaka commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,834 @@
+   Compress *data* (a :term:`bytes-like object`), returning the compressed
+   data as a :class:`bytes` object.
+
+   The *level* argument is an int object controlling the level of
```
Why "int object"? Why not simply "integer"?

There are many other occurrences.

emmatyping replied:
I switched to "integer" to use simple language.

serhiy-storchaka commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,834 @@
+   decompression, passing a digested dictionary will reduce the overhead of
+   loading the dictionary.
+
+    .. list-table:: Difference for compression
```
I think that simple nested item lists would look better here. Maybe not even nested item lists, just an item lists with two or three paragraphs in every item.

emmatyping replied:
Sorry, I'm not sure I understand what you have in mind, perhaps it can be resolved in a follow-up?

serhiy-storchaka commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,834 @@
+      inputs by finding large matches at greater distances. It increases memory
+      usage and window size.
+
+      ``True`` or ``0`` enable long distance matching while ``False`` or ``1``
```
This is VERY counterintuitive and error prone.

emmatyping replied:
Fortunately it is a typo :)

Thanks for catching this

emmatyping replied:
I fixed the 3 instances of this.

serhiy-storchaka commented on Doc/library/compression.zstd.rst:
```diff
@@ -0,0 +1,834 @@
+      the checksum. If there is a mismatch a :class:`ZstdError` exception is
+      raised.
+
+      ``True`` or ``0`` enable checksum generation while ``False`` or ``1``
```
Again!

serhiy-storchaka reviewed (approved):
Besides some nitpicks, LGTM.

emmatyping committed:
Expand on the connection between level and compression_level

emmatyping committed:
Resolve review suggestions

- Use "integer" instead of "int object"
- Use > 0 rather than >=1 for nb_workers
